source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc

```diff
@@ -165,6 +168,69 @@ void UpstreamProxyProtocolSocketFactory::hashKey(
   }
 }
 
+std::unordered_map<uint8_t, std::vector<unsigned char>> UpstreamProxyProtocolSocket::buildCustomTLVs() {
+  std::unordered_map<uint8_t, std::vector<unsigned char>> custom_tlvs;
+
+  const auto& upstream_info = callbacks_->connection().streamInfo().upstreamInfo();
+  if (upstream_info == nullptr) {
+    return custom_tlvs;
+  }
+
+  Upstream::HostDescriptionConstSharedPtr host = upstream_info->upstreamHost();
+  if (host == nullptr) {
+    return custom_tlvs;
+  }
+
+  auto metadata = host->metadata();
+  if (metadata == nullptr) {
+    return custom_tlvs;
+  }
+
+  const auto& filter_metadata = metadata->filter_metadata();
+  const auto filter_it = filter_metadata.find("envoy.transport_sockets.proxy_protocol");
+  if (filter_it == filter_metadata.end()) {
+    ENVOY_LOG(trace, "no custom TLVs found in upstream host metadata");
+    return custom_tlvs;
+  }
+
+  // Construct the map from the metadata.
+  const auto& metadata_struct = filter_it->second;
+  for (const auto& field : metadata_struct.fields()) {
+    const std::string& key = field.first;
+    const auto& value = field.second;
+
+    // Assume the metadata key is the TLV type in hex format (e.g., "0xD3").
+    uint8_t tlv_type;
+    try {
+      tlv_type = static_cast<uint8_t>(std::stoi(key, nullptr, 16));
```

Comment:
Addressed in https://github.com/timflannagan/envoy/pull/1/commits/fe77f985401b45fdd9b349b8f0dd73c5a9c5115d. I ended up embedding that type in the PP2 upstream transport socket proto, but it's a bit awkwardly placed since 1.) you can't set this config field in the upstream_proxy_protocol transport socket config, and 2.) I'm not sure we want to support populating custom TLVs from filter metadata and reading from the transport socket config.